sdk/macro-frozen-abi/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+extern crate proc_macro;
+
+#[macro_use]
+extern crate lazy_static;
+
+lazy_static! {
+    static ref SPECIALIZATION_DETECTOR_INJECTED: std::sync::atomic::AtomicBool =
+        std::sync::atomic::AtomicBool::new(false);
+}
+
+// This file littered with these essential cfgs so ensure them.
+#[cfg(not(any(RUSTC_WITH_SPECIALIZATION, RUSTC_WITHOUT_SPECIALIZATION)))]
+compile_error!("rustc_version is missing in build dependency and build.rs is not specified");
+
+#[cfg(any(RUSTC_WITH_SPECIALIZATION, RUSTC_WITHOUT_SPECIALIZATION))]
+use proc_macro::TokenStream;
+
+// Define dummy macro_attribute and macro_derive for stable rustc
+
+#[cfg(RUSTC_WITHOUT_SPECIALIZATION)]
+#[proc_macro_attribute]
+pub fn frozen_abi(_attrs: TokenStream, item: TokenStream) -> TokenStream {
+    item
+}
+
+#[cfg(RUSTC_WITHOUT_SPECIALIZATION)]
+#[proc_macro_derive(AbiSample)]
+pub fn derive_abi_sample(_item: TokenStream) -> TokenStream {
+    "".parse().unwrap()
+}
+
+#[cfg(RUSTC_WITHOUT_SPECIALIZATION)]
+#[proc_macro_derive(AbiDigest)]
+pub fn derive_abi_digest(_item: TokenStream) -> TokenStream {
+    "".parse().unwrap()
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+use proc_macro2::{Span, TokenStream as TokenStream2, TokenTree::Group};
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+use quote::quote;
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+use syn::{
+    parse_macro_input, Attribute, AttributeArgs, Fields, Ident, Item, ItemEnum, ItemStruct,
+    ItemType, Lit, Meta, NestedMeta, Variant,
+};
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+fn filter_serde_attrs(attrs: &[Attribute]) -> bool {
+    let mut skip = false;
+
+    for attr in attrs {
+        let ss = &attr.path.segments.first().unwrap().ident.to_string();
+        if ss.starts_with("serde") {
+            for token in attr.tokens.clone() {
+                if let Group(token) = token {
+                    for ident in token.stream() {
+                        if ident.to_string() == "skip" {
+                            skip = true;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    skip
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+fn filter_allow_attrs(attrs: &mut Vec<Attribute>) {
+    attrs.retain(|attr| {
+        let ss = &attr.path.segments.first().unwrap().ident.to_string();
+        ss.starts_with("allow")
+    });
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+fn quote_for_specialization_detection() -> TokenStream2 {
+    if !SPECIALIZATION_DETECTOR_INJECTED.load(std::sync::atomic::Ordering::Relaxed) {
+        SPECIALIZATION_DETECTOR_INJECTED.store(true, std::sync::atomic::Ordering::Relaxed);
+        quote! {
+            mod specialization_detector {
```

Comment:
I couldn't come up with a more direct elegant solution like [this](https://github.com/solana-labs/solana/pull/8012/files#diff-c60af798770ee7e49103563257a0f1c9R13).